File: python/riot_sdk/facade/order_client.py
from __future__ import annotations

from typing import TYPE_CHECKING, Any

from kiota_abstractions.base_request_configuration import RequestConfiguration

from riot_sdk.core.business_response import ensure_success, require_response
from riot_sdk.core.exceptions import RiotApiException
from riot_sdk.core.order_ref import OrderRef
from riot_sdk.generated.order.models.mission_d_t_o import MissionDTO
from riot_sdk.generated.order.models.order_record_d_t_o_object import OrderRecordDTOObject

if TYPE_CHECKING:
    from riot_sdk.facade.session import RiotSession


def _to_order_ref(result: Any, source: str) -> OrderRef:
    if (
        result is None
        or result.id is None
        or not result.order_id
        or not result.upper_id
        or result.order_state is None
    ):
        raise RiotApiException(
            f"{source} returned incomplete order identifiers.",
            business_code="order-ref-missing",
        )
    return OrderRef(
        id=int(result.id),
        order_id=result.order_id,
        upper_id=result.upper_id,
        order_state=int(result.order_state),
    )


class OrderClient:
    """Thin order-module facade."""

    def __init__(self, session: RiotSession) -> None:
        self._session = session

    async def create_move_order(
        self,
        upper_id: str,
        appoint_vehicle_key: str,
        map_id: int,
        destination_station_id: int,
        order_name: str | None = None,
    ) -> OrderRef:
        """POST /api/order/v1/add/byDefaultMissions — single-segment move (BC-ORDER-001)."""
        if not upper_id or not upper_id.strip():
            raise ValueError("upper_id is required")
        if not appoint_vehicle_key or not appoint_vehicle_key.strip():
            raise ValueError("appoint_vehicle_key is required")

        client = self._session.create_generated_order_client()
        body = OrderRecordDTOObject(
            appoint_vehicle_key=appoint_vehicle_key,
            is_appoint_enable=1,
            lock_status=0,
            order_name=order_name.strip() if order_name and order_name.strip() else upper_id,
            upper_id=upper_id,
            mission=[
                MissionDTO(
                    type="move",
                    map_id=map_id,
                    destination=destination_station_id,
                )
            ],
        )
        response = require_response(
            await client.api.order.v1.add.by_default_missions.post(body),
            "byDefaultMissions",
        )
        ensure_success(response.code, response.message)
        return _to_order_ref(response.result, "byDefaultMissions")

    async def get_order_by_upper_id(self, upper_id: str) -> OrderRef:
        """GET /api/order/v1/orderRecord/detailByUpperId/{upperId} (BC-ORDER-005)."""
        if not upper_id or not upper_id.strip():
            raise ValueError("upper_id is required")

        client = self._session.create_generated_order_client()
        response = require_response(
            await client.api.order.v1.order_record.detail_by_upper_id.by_upper_id(
                upper_id
            ).get(),
            "detailByUpperId",
        )
        ensure_success(response.code, response.message)
        return _to_order_ref(response.result, "detailByUpperId")

    async def get_order_by_order_id(self, order_id: str) -> OrderRef:
        """GET /api/order/v1/orderRecord/detailByOrderId/{orderId} (BC-ORDER-005)."""
        if not order_id or not order_id.strip():
            raise ValueError("order_id is required")

        client = self._session.create_generated_order_client()
        response = require_response(
            await client.api.order.v1.order_record.detail_by_order_id.by_order_id(
                order_id
            ).get(),
            "detailByOrderId",
        )
        ensure_success(response.code, response.message)
        return _to_order_ref(response.result, "detailByOrderId")

    async def priority_exec(self, order_id: str) -> None:
        """POST /api/order/v1/orderRecordPriorityExec?orderTaskKey={orderId} (BC-ORDER-014)."""
        if not order_id or not order_id.strip():
            raise ValueError("order_id is required")

        client = self._session.create_generated_order_client()
        request = client.api.order.v1.order_record_priority_exec
        query = request.OrderRecordPriorityExecRequestBuilderPostQueryParameters()
        query.order_task_key = order_id
        config = RequestConfiguration(query_parameters=query)
        response = require_response(
            await request.post(request_configuration=config),
            "orderRecordPriorityExec",
        )
        ensure_success(response.code, response.message)

    @property
    def raw(self):
        """Underlying Kiota client for endpoints not yet wrapped."""
        return self._session.create_generated_order_client()

File: csharp/RIoT.Sdk.Facade/OrderClient.cs
using RIoT.Sdk.Core;

namespace RIoT.Sdk.Facade;

/// <summary>
/// Thin order-module facade. Expose additional endpoints as needed.
/// </summary>
public sealed class OrderClient
{
    private readonly RiotSession _session;

    internal OrderClient(RiotSession session) => _session = session;

    /// <summary>
    /// POST /api/order/v1/add/byDefaultMissions — single-segment move order (BC-ORDER-001).
    /// Throws <see cref="RiotApiException"/> on business failure (ADR-sdk-0005).
    /// </summary>
    public async Task<OrderRef> CreateMoveOrderAsync(
        string upperId,
        string appointVehicleKey,
        int mapId,
        int destinationStationId,
        string? orderName = null,
        CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(upperId);
        ArgumentException.ThrowIfNullOrWhiteSpace(appointVehicleKey);

        var client = _session.CreateGeneratedOrderClient();
        var body = new RIoT.Sdk.Generated.Order.Models.OrderRecordDTOObject
        {
            AppointVehicleKey = appointVehicleKey,
            IsAppointEnable = 1,
            LockStatus = 0,
            OrderName = string.IsNullOrWhiteSpace(orderName) ? upperId : orderName,
            UpperId = upperId,
            Mission =
            [
                new RIoT.Sdk.Generated.Order.Models.MissionDTO
                {
                    Type = "move",
                    MapId = mapId,
                    Destination = destinationStationId,
                },
            ],
        };

        var response = RiotBusinessResponse.RequireResponse(
            await client.Api.Order.V1.Add.ByDefaultMissions
                .PostAsync(body, cancellationToken: cancellationToken)
                .ConfigureAwait(false),
            "byDefaultMissions");

        RiotBusinessResponse.EnsureSuccess(response.Code, response.Message);
        return ToOrderRef(response.Result, "byDefaultMissions");
    }

    /// <summary>
    /// GET /api/order/v1/orderRecord/detailByUpperId/{upperId} (BC-ORDER-005).
    /// Throws <see cref="RiotApiException"/> on business failure (ADR-sdk-0005).
    /// </summary>
    public async Task<OrderRef> GetOrderByUpperIdAsync(
        string upperId,
        CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(upperId);

        var client = _session.CreateGeneratedOrderClient();
        var response = RiotBusinessResponse.RequireResponse(
            await client.Api.Order.V1.OrderRecord.DetailByUpperId[upperId]
                .GetAsync(cancellationToken: cancellationToken)
                .ConfigureAwait(false),
            "detailByUpperId");

        RiotBusinessResponse.EnsureSuccess(response.Code, response.Message);
        return ToOrderRef(response.Result, "detailByUpperId");
    }

    /// <summary>
    /// GET /api/order/v1/orderRecord/detailByOrderId/{orderId} (BC-ORDER-005).
    /// Throws <see cref="RiotApiException"/> on business failure (ADR-sdk-0005).
    /// </summary>
    public async Task<OrderRef> GetOrderByOrderIdAsync(
        string orderId,
        CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(orderId);

        var client = _session.CreateGeneratedOrderClient();
        var response = RiotBusinessResponse.RequireResponse(
            await client.Api.Order.V1.OrderRecord.DetailByOrderId[orderId]
                .GetAsync(cancellationToken: cancellationToken)
                .ConfigureAwait(false),
            "detailByOrderId");

        RiotBusinessResponse.EnsureSuccess(response.Code, response.Message);
        return ToOrderRef(response.Result, "detailByOrderId");
    }

    /// <summary>
    /// POST /api/order/v1/orderRecordPriorityExec?orderTaskKey={orderId} (BC-ORDER-014).
    /// Promotes a QUEUEING order ahead of other QUEUEING orders on the same vehicle.
    /// Pass the string orderId (not numeric id / upperId).
    /// </summary>
    public async Task PriorityExecAsync(
        string orderId,
        CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(orderId);

        var client = _session.CreateGeneratedOrderClient();
        var response = RiotBusinessResponse.RequireResponse(
            await client.Api.Order.V1.OrderRecordPriorityExec
                .PostAsync(config =>
                {
                    config.QueryParameters.OrderTaskKey = orderId;
                }, cancellationToken)
                .ConfigureAwait(false),
            "orderRecordPriorityExec");

        RiotBusinessResponse.EnsureSuccess(response.Code, response.Message);
    }

    /// <summary>
    /// Access the underlying Kiota client for endpoints not yet wrapped.
    /// </summary>
    public RIoT.Sdk.Generated.Order.OrderClient Raw => _session.CreateGeneratedOrderClient();

    private static OrderRef ToOrderRef(
        RIoT.Sdk.Generated.Order.Models.OrderRecordObject? result,
        string source)
    {
        if (result?.Id is null
            || string.IsNullOrWhiteSpace(result.OrderId)
            || string.IsNullOrWhiteSpace(result.UpperId)
            || result.OrderState is null)
        {
            throw new RiotApiException(
                $"{source} returned incomplete order identifiers.",
                businessCode: "order-ref-missing");
        }

        return new OrderRef(
            result.Id.Value,
            result.OrderId,
            result.UpperId,
            result.OrderState.Value);
    }
}
